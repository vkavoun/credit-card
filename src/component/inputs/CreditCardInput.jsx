import React, { useState } from 'react';
import PropTypes from 'prop-types';

const maskCardNumber = value => value.replace(/(\d{4})/g, 'xxxx ').replace(/(^\s+|\s+$)/, '');

const formatCardNumber = value => value.replace(/(\d{4})/g, '$1 ').replace(/(^\s+|\s+$)/, '');

export default function CreditCardInput({ cardNumber, handleCardNumberInput }) {
  const [isCardNumberMasked, setIsCardNumberMasked] = useState(true);

  const toggleCardNumberMask = () => {
    setIsCardNumberMasked(!isCardNumberMasked);
  };

  return (
    <div key="credit-card-number-wrapper">
      <label>Credit Card Number*</label>
      <input
        id="card-keyboard-manual-entry__keyed-entry-input"
        name="card-keyboard-manual-entry__keyed-entry-input"
        data-testid="card-number"
        autoFocus
        type="tel"
        value={isCardNumberMasked ? maskCardNumber(cardNumber) : formatCardNumber(cardNumber)}
        onChange={e => handleCardNumberInput(e)}
        onFocus={() => toggleCardNumberMask()}
        onBlur={() => toggleCardNumberMask()}
      />
    </div>
  );
}

CreditCardInput.propTypes = {
  cardNumber: PropTypes.string.isRequired,
  handleCardNumberInput: PropTypes.func.isRequired,
};
