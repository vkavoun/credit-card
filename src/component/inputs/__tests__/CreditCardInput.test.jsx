import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import CreditCardInput from '../CreditCardInput';

describe('CreditCardInput', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<CreditCardInput cardNumber={''} handleCardNumberInput={jest.fn()} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<CreditCardInput cardNumber={''} handleCardNumberInput={jest.fn()} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
