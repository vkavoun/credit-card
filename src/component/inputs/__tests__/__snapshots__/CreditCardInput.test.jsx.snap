// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardInput renders correctly 1`] = `
<div>
  <label>
    Credit Card Number*
  </label>
  <input
    autoFocus={true}
    data-testid="card-number"
    id="card-keyboard-manual-entry__keyed-entry-input"
    name="card-keyboard-manual-entry__keyed-entry-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="tel"
    value=""
  />
</div>
`;
